cmd/snap-bootstrap/cmd_initramfs_mounts_cvm.go

```diff
@@ -44,9 +47,173 @@ func (*genericCVMModel) Grade() asserts.ModelGrade {
+		}
+
+		pm.GptLabel = p.GptLabel
+
+		// All detected partitions are mounted by default under /run/mnt/<GptLabel of partition>
+		pm.Where = filepath.Join(boot.InitramfsRunMntDir, p.GptLabel)
+
+		if p.ReadOnly {
+			// XXX: currently only a single read-only partition/overlay fs lowerdir is permitted.
+			if foundReadOnlyPartition != "" {
+				return nil, errors.New("manifest contains multiple partitions marked as read-only")
+
+			}
+			foundReadOnlyPartition = pm.GptLabel
+
+			// systemd-mount will run fsck by default when attempting to mount the partition and potentially corrupt it.
+			// This will cause dm-verity to fail when attempting to set up the dm-verity mount.
+			// fsck should be/is run by the encrypt-cloud-image tool prior to generating dm-verity data.
+			pm.Opts.NeedsFsck = false
+			pm.Opts.VerityRootHash = p.RootHash
+
+			// Auto-discover verity device from disk.
+			verityPartition, err := disk.FindMatchingPartitionWithPartLabel(p.GptLabel + "-verity")
+			if err != nil {
+				return []partitionMount{}, err
+			}
+			pm.Opts.VerityHashDevice = verityPartition.KernelDeviceNode
+		} else {
+			// Only one writable partition is permitted.
+			if foundWritablePartition != "" {
+				return nil, errors.New("manifest contains multiple non read-only partitions")
+			}
+			// Manifest contains a partition meant to be used as a writable overlay for the non-ephemeral vm case.
+			// If it is encrypted, its key will be autodiscovered based on its FsLabel later.
+			foundWritablePartition = p.GptLabel
+			pm.Opts.NeedsFsck = true
+		}
+
+		partitionMounts = append(partitionMounts, pm)
+	}
+
+	if foundReadOnlyPartition == "" {
+		return nil, errors.New("manifest doesn't contain any partition marked as read-only")
+	}
+
+	// If no writable partitions were found in the manifest, Configure a tmpfs filesystem for the upper and workdir layers
+	// of the final rootfs mount.
+	if foundWritablePartition == "" {
+		foundWritablePartition = "writable-tmp"
+
+		pm := partitionMount{
+			Where:    filepath.Join(boot.InitramfsRunMntDir, "writable-tmp"),
+			GptLabel: "writable-tmp",
+			Opts: &systemdMountOptions{
+				Tmpfs: true,
+			},
+		}
+
+		partitionMounts = append(partitionMounts, pm)
+	}
+
+	// Configure the merged overlay filesystem mount.
+	pm := partitionMount{
+		Where:    boot.InitramfsDataDir,
+		GptLabel: "cloudimg-rootfs",
+		Opts: &systemdMountOptions{
+			Overlayfs: true,
+			LowerDirs: []string{filepath.Join(boot.InitramfsRunMntDir, foundReadOnlyPartition)},
+			UpperDir:  filepath.Join(boot.InitramfsRunMntDir, foundWritablePartition, "upper"),
+			WorkDir:   filepath.Join(boot.InitramfsRunMntDir, foundWritablePartition, "work"),
+		},
+	}
+
+	partitionMounts = append(partitionMounts, pm)
+
+	return partitionMounts, nil
+}
+
+var createOverlayDirs = func(path string) error {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	if err := os.Mkdir(path, fi.Mode()); err != nil && !os.IsExist(err) {
```

Comment:
won't we always get an IsExist error, otherwise how can the previous os.Stat work?

Reply:
I reworked that bit.